src/core/ipe-setup.c

```diff
@@ -0,0 +1,110 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "copy.h"
+#include "dirent-util.h"
+#include "fd-util.h"
+#include "fileio.h"
+#include "ipe-setup.h"
+#include "path-util.h"
+
+#define IPE_SECFS_DIR "/sys/kernel/security/ipe"
+#define IPE_SECFS_NEW_POLICY IPE_SECFS_DIR "/new_policy"
+#define IPE_SECFS_POLICIES IPE_SECFS_DIR "/policies/"
+#define IPE_INPUT_POLICY_DIR "/etc/ipe"
+
+int ipe_setup(void) {
+#if ENABLE_IPE
+        _cleanup_closedir_ DIR *policy_dir = NULL;
+        int r;
+
+        /* Very quick smoke tests first: this is in the citical, sequential boot path, and in most cases it
+         * is unlikely this will be configured, so do the fastest existence checks first and immediately
+         * return if there's nothing to do. */
+
+        if (access(IPE_SECFS_DIR, F_OK) < 0) {
+                log_debug_errno(errno, "IPE support is disabled in the kernel, ignoring: %m");
+                return 0;
+        }
+
+        if (access(IPE_INPUT_POLICY_DIR, F_OK) < 0) {
+                log_debug_errno(errno,
+                                "No IPE custom policy directory %s, ignoring: %m",
+                                IPE_INPUT_POLICY_DIR);
+                return 0;
+        }
+
+        policy_dir = opendir(IPE_INPUT_POLICY_DIR);
+        if (!policy_dir)
+                return log_error_errno(
+                                errno,
+                                "Failed to open the IPE custom policy directory %s: %m",
+                                IPE_INPUT_POLICY_DIR);
+
+        FOREACH_DIRENT(de, policy_dir, return log_error_errno(errno, "Failed to read %s directory: %m", IPE_INPUT_POLICY_DIR)) {
+                _cleanup_free_ char *policy_name = NULL, *output_path = NULL, *activate_path = NULL;
+                _cleanup_close_ int input = -EBADF, output = -EBADF;
+                const char *suffix;
+
+                if (de->d_type != DT_REG)
+                        continue;
+
+                suffix = endswith(de->d_name, ".p7b");
+                if (!suffix)
+                        continue;
+
+                policy_name = strndup(de->d_name, suffix - de->d_name);
+                if (!policy_name)
+                        return log_oom();
+
```

Comment:
one more safety check: add "filename_is_valid(policy_name)" check here. Just as paranoia, since otherwise files named just `.p7b` or `...p7b` will trigger quite confusing behaviour.